### DISCUSSION TITLE
Changing the `start` command to run `auto-build` by default

### DISCUSSION THREAD

**Discussion by pedroigor** [keycloak/keycloak#10451] [Archive]
Based on the feedback we have received so far from the community (and really helpful feedback), one of the main problems people are struggling with is whether you should run the `build` command before actually running the `start` command to start the server.

This is also related to the confusion around what is a *build option* and what is a *configuration option*.

In order to avoid confusion when *starting* the server, people should just run `start` without worrying about whether an option is targeted for the `build` or the `start` command. But just rely on the server to perform an automatic build whenever a build option change.

Right now, people should be able to achieve that by running the server using the `--auto-build` option:

```
kc.sh start --auto-build --db=postgres --cache=local <any-option>
```

But our guides and documentation are using the `build` command whenever a build option is being documented (e.g.: database, clustering, etc) instead of just using the `start` command with the `--auto-build` option. That causes unnecessary confusion if you just want to start the server using any of the options available (no matter if they are build/runtime) and the performance penalty on the start-up is not a problem for you.

That said, the proposal here is to:

* Set the `--auto-build` option by default when running the `start` command, similarly to when running `start-dev` but the command is still targeted for production.
* Change guides accordingly to just use the `start` command when documenting how to use a specific build option
* Introduce a `--no-auto-build` option that when used with the `start` command it is going to skip the auto-build/re-augmentation step completly and just start the server based on any configuration previously set using the `build` command.
* Review, update, and highlight, the importance of using the `--no-auto-build` option and using the `build` command directly in order to achieve the best startup time and memory footprint.

In the end, we consider the optimizations and performance improvements brought by Quarkus and `build` command as a second citizen (not in terms of importance) and not as a primary/mandatory concern as we are doing today as well as in favor of a better UX.

Please, let us know what you think about this proposal and if it helps to make things easier when configuring/starting the server.

**Comment by dasniko** [comment]
The more I think about it, the more I can agree with this approach.

For me personally, it's not a big deal to handle `build`, `start` and `--auto-build`, but I won't consider me as a regular user. I'm working with the Quarkus-based distro now since late summer '21, additionally I'm working with Quarkus and I know about all the importance of built-time optimizations. But, again, that's personally me.

The most confusion I see at customers is about the terms, that one has to _build_ a Keycloak _image_ before it can be started in production. Most people assume with _build_ a Java/Maven build, producing all the bytecode (not only the KC optimizations) and with _image_ something like a container image and don't know why they have to do it, as they want to deploy KC in an old-fashioned way.

So, having an automatic build when just using `start` will be a good idea, IMHO. And then, pointing out, that there is a performance optimization feature if you `build` the KC image beforehand.

**Comment by stianst** [comment]
I think this is the right way to go as we've spent a long time and many iterations trying to make the build/start experience as smooth as possible, but it's just not ever going to get there, so having the build optimisation as an opt-in rather an opt-out experience makes a lot of sense to me.

To add to what @pedroigor has written I would also have optimising the build centralised in a single guide (Optimising the server runtime guide or something like that), and not dilute other guides with this concept at all.

**Comment by stianst** [comment]
For a migration perspective from Keycloak 17 we need to list in the migration guide the change to make everyone aware that an auto-build is now always happening unless "--no-auto-build" is used. Maybe we should also consider keeping "--auto-build" for a little while, but it would print a message saying that it is deprecated/no longer used.

**Comment by stianst** [comment]
Not sure what is the best name for the option. "--no-auto-build" doesn't seem quite right, I'd rather have something that says "use the optimised build"; --use-build, --optimised-build, --optimised, dunno..

**Comment by dasniko** [comment]
Any update on this, @pedroigor?
Will KC18 bring this change in default startup behavior?

**Reply by pedroigor** [reply]
Yeah, it is targeted for 18.

**Reply by beniblanco78** [reply]
Hello,
We have dedicated a script to start the keycloak server that was based on the existing distribution.
Now that we move to Quarku distrib we need to use the new commands build (we understand that the build process has to be ran once) and start separately (for performance reason) 
The question is now how can we identify in our start up script if the build process has already been done or not?
Also another question : if we are using a database storage how can we perform a cleanup of the database?

thanks in advance,
Nicolas

**Reply by beniblanco78** [reply]
Also got a quick question regarding build and start.
Which options are mandatory to be kept in start mode?
Right now we are using following command (and we want to separate build and start ) : 
$KC_OPTS = "--log=file --log-file=$KeycloakLogFile --log-level=debug --db-url jdbc:oracle:thin:@xxx --db-username xxx --db-password xxx 
cmd /C "$PSScriptRoot\bin\kc.bat --verbose start --auto-build --import-realm  --http-enabled=true --health-enabled=true --metrics-enabled=true --http-relative-path /auth --spi-user-storage-provider=xxx --hostname-strict=false --cache=local $KC_OPTS"
